피그마 3번 페이지를 개발

kong33 opened the pull request:
피그마 3번 페이지를 개발했습니다.

choosing component를 추가해 페이지를 개발했습니다.
font가 전부 google font소스임을 확인한 후 인터넷 url로 폰트 연결하는 코드를 index.html 헤더에 작성했습니다.
starting 페이지에서 choosing 페이지를 라우팅하는 코드를 추가했습니다.

kong33 committed:
make choosing page

make choosing page. add font code

kong33 committed:
Merge branch 'main' into ChoosePage

kong33 committed:
eslint 적용

kong33 committed:
Merge branch 'ChoosePage' of https://github.com/kong33/kiosk into ChoosePage

kong33 committed:
./

ENvironmentSet commented:
resolves: #5 

이렇게 이슈를 연결해서 pr을 병합할 때 같이 닫히게 할 수 있어요! #7 참고 부탁드릴게요!

ENvironmentSet commented on src/App.jsx:
```diff
@@ -1,10 +1,11 @@
-import { Redirect, Route, Switch } from 'wouter'
-import { Starting } from './pages/Starting'
-
+import { Redirect, Route, Switch } from 'wouter';
```
ESLint 설정할 때 세미콜론을 사용하지 않는 것으로 코딩 컨벤션을 정의했던 거 같은데 제대로 안 됐나 봐요. 이건 제가 다른 이슈로 만들어서 해결해볼게요.

kong33 replied:
앗 넵 그러면 앞으로 세미콜론 작성 안 하면 될까요??

ENvironmentSet replied:
네!

ENvironmentSet commented on src/App.jsx:
```diff
@@ -1,10 +1,11 @@
-    <Route path='/'> <Starting /> </Route>
-
-    <Route><Redirect to='/' /> {/** Or should we redirect to 404 page? */} </Route>
+    <Route exact path='/starting'> <Starting /> </Route>
```
`Starting` 페이지의 `path`는 `/`가 맞아요! 제가 이전에 #2를 병합하기 전에 `/starting`에서 `/`로 바꿨었어요~

ENvironmentSet commented on src/App.jsx:
```diff
@@ -1,10 +1,11 @@
-    <Route><Redirect to='/' /> {/** Or should we redirect to 404 page? */} </Route>
+    <Route exact path='/starting'> <Starting /> </Route>
+    <Route path='/choosing'><Choosing /></Route>
+    <Route><Redirect to='/starting' /> {/** Or should we redirect to 404 page? */}</Route>
```
여기도 같이 수정 부탁드릴게요!

ENvironmentSet commented on src/components/Button/index.jsx:
```diff
@@ -0,0 +1,32 @@
+import { css } from '@emotion/react';
+import {Link} from 'wouter';
```
불러올 요소의 나열이랑 중괄호 두 개는 떨어뜨려 놓는 게 보기 더 편하다 생각해요. 1번 줄 처럼요! 가은님 생각은 어떠실까요?

kong33 replied:
저도 띄워서 두는 게 편할 것 같아요! 앞으로 일괄적으로 띄어서 작성하겠습니당!

ENvironmentSet commented on src/components/Button/index.jsx:
```diff
@@ -0,0 +1,32 @@
+import {Link} from 'wouter';
+
+
+export const Button = ({buttonText, to, backgroundColor, opacity}) =>{
```
여기도 2번 줄 리뷰랑 비슷해요. JS에서 객체 리터럴이나 비구조화 할당 패턴, import 문을 사용할 때에는 중괄호와 이름들을 떨어뜨려 놓는 게 일반적인 컨벤션입니다. 그게 더 보기 편해요~

ENvironmentSet replied:
`=>` 이후에도 빈 공간이 필요해 보여요.

ENvironmentSet replied:
Button 컴포넌트로 그려지는 엘리먼트의 크기를 조정하거나 글자 색을 바꾼다거나 하는 추가 스타일 적용을 지원하기 위해 `className` prop을 지원해주시는 건 어떨까요?

ENvironmentSet replied:
default 값을 `??`나 `||`로 설정하는 것도 좋지만, 아래와 같이 매개변수를 받는 부분에서 설정할 수도 있어요.

```js
const Foo = ({ a = 1 }) => a

console.log(Foo({}))
```

kong33 replied:
className prop 지원하게 변경하겠습니다!
매개변수를 받을 때 설정하면 좋은 점은 가독성인가요??

kong33 replied:
className으로 따로 css 지정하지 않으면 기본 스타일을 유지할 수 있게 코드를 짜는 게 좋을까요?

kong33 replied:
그리고 className으로 받으면 default 값 굳이 위의 방법대로 설정 안 하고
` const buttonStyles =css
    background-color: '#000'; 
    opacity:0.5;
    display: flex;
    width: 648px;
    height: 132px;
    justify-content: center;
    align-items: center;
    flex-shrink: 0;
    border-radius: 15px;
    margin:20px;
    font-size: 30px;
    font-family: 'Noto Sans KR', sans-serif;
    border:none;   
    `

이렇게 작성해둔 후 나중에 css important로 작성하는 건 어떨까요?@

ENvironmentSet replied:
네. 변수의 값을 읽을 때 값이 없으면 미리 정의된 값을 사용하는 동작보다는 변수의 값이 설정 안 되어 있을 때 미리 정의된 값으로 설정하는 동작이 더 기본값이라는 개념을 충실하게 구현한다 생각해요. 기본값을 단 한 번만 적어주면 돼서 기본값을 까먹고 안 적는다거나, 잘못 적는 문제도 없고 혹시나 기본값이 사용처마다 달라져서 개념적인 불일치가 생기는 일도 없지요.

ENvironmentSet replied:
네, className으로 들어오는 css는 추가 스타일인지라 기본 스타일이 별도로 존재하는 게 코드 재사용 측면이나 사용성 측면에서 좋습니다. 더해서 css important는 가급적 피하고 싶은 게, 한 번 사용하기 시작하면 그 important rule을 덮어쓰기 위해 더 쓰게 되고 그러다 보면 우선순위가 완전히 꼬여서 css 코드를 관리하기가 어려워질 거 같거든요... 원래 하시던 대로 컴포넌트 안에 기본 스타일을 정의해주시는 게 더 좋을 거 같습니다.

kong33 replied:
넵 알겠습니다!!

ENvironmentSet commented on src/components/Button/index.jsx:
```diff
@@ -0,0 +1,32 @@
+export const Button = ({buttonText, to, backgroundColor, opacity}) =>{
+    
+  const buttonSyles =css`
+    background-color: ${backgroundColor || '#000'}; 
```
최신 버전의 JS에서는 `||`보다 더 안전한 `??`가 있어요. 거의 비슷한데 `??`는 `undefined`와 `null`에 대해서만 바로 뒤 코드를 계산해 결과로 내놓는다는 차이점을 같습니다.

ENvironmentSet replied:
연산자나 변수 선언의 `=` 앞뒤로도 공백을 넣어주시는 게 좋아요. `a=1`보다 `a = 1`이 더 보기 편하지요.

kong33 replied:
넵 수정하겠습니다!

ENvironmentSet commented on src/components/Button/index.jsx:
```diff
@@ -0,0 +1,32 @@
+    `
+    
+  return(
+    <div>
```
`div`가 꼭 사용되어야 할 이유가 있었을까요? 제가 보기에는 불필요해 보여요..!

kong33 replied:
넵 저도 필요 없어 보여서 삭제했습니다!

ENvironmentSet commented on src/components/Button/index.jsx:
```diff
@@ -0,0 +1,32 @@
+    <div>
+      <button
+        css={buttonSyles}>
+        <Link to={to} css={css`color:#FFF; text-decoration: none;`}>{buttonText}</Link>
```
모든 버튼이 `Link`로 구현되어야 할까요? 이 부분은 더 추상화해도 좋을 거 같아요.

ENvironmentSet commented on src/components/Button/index.jsx:
```diff
@@ -0,0 +1,32 @@
+    border-radius: 15px;
+    margin:20px;
+    font-size: 30px;
+    font-family: 'Noto Sans KR', sans-serif;
```
이 폰트는 버튼 내부 문장에 대해서만 적용되어야 하는 폰트인가요? 그렇지 않다면 전역적으로 폰트를 설정할 수 있을 거 같아요.

kong33 replied:
전역 폰트 설정은 밑에 말씀하신 ko ,en 처럼 설정하는 걸  말씀하신거겠죠??

ENvironmentSet replied:
그것보다는 #7 처럼 하는 방법을 말씀드린 거였어요! 직접 하나하나 조정하는 것 보다 정교하지는 않지만 아직까지 하나하나 조정해야 하는 경우를 본 적 없으니 이렇게 해도 좋을 거 같아서요~

kong33 replied:
아 넵넵! 그러면 영문 폰트가 font에 있는 폰트 이외에 하나 더 추가된 것 같은데 그것만 폰트 파일에 추가해두면 될까요??

ENvironmentSet replied:
네네 부탁드릴게요~

ENvironmentSet commented on src/components/Button/index.jsx:
```diff
@@ -0,0 +1,32 @@
+    margin:20px;
+    font-size: 30px;
+    font-family: 'Noto Sans KR', sans-serif;
+    border:none;
```
`border-radius`나 `border: none` 둘 중 하나는 없어도 되는 코드로 보여요.

kong33 replied:
border-radius는 둥글게 한 속성이고 border: none은 button 태그를 쓰다 보니 테두리에 음영이 생겨서 없앤거라 다른 역할을 하는데 혹시 이견 있으실까요?!

ENvironmentSet replied:
아~ 그런 차이가 있군요~ 제가 테스트 해봤을 때에는 하나만 있어도 충분한 것 처럼 보였는데 제가 잘못 봤나 봐요. 확인해주셔서 감사합니다~

ENvironmentSet commented on src/components/Button/index.jsx:
```diff
@@ -0,0 +1,32 @@
+    background-color: ${backgroundColor || '#000'}; 
+    opacity: ${opacity || 0.5};
+    display: flex;
+    width: 648px;
```
범용 컴포넌트들은 다양한 사용처에서 다양한 크기와 형태로 사용될 텐데, 굳이 사이즈를 정해 주신 이유가 있으실까요?

kong33 replied:
제가 봤을 때 사이즈가 거의 비슷해서 우선 정해두었는데 혹시 그 key, place, scent에도 이 버튼 사용하실 거면 그건 사이즈나 스타일이 많이 달라서 굳이 정할 필요 없을 것 같아요! 어떻게 생각하세요??

ENvironmentSet replied:
아! 그런 거라면 그냥 두셔도 될 거 같습니다~

ENvironmentSet commented on src/pages/Choosing/index.jsx:
```diff
@@ -0,0 +1,109 @@
+  return(
+    <div 
+      css={css`
+    width: 100%;
```
들여쓰기 확인 부탁드려요.

ENvironmentSet commented on src/pages/Choosing/index.jsx:
```diff
@@ -0,0 +1,109 @@
+import { css } from '@emotion/react'
+import {Button} from '../../components/Button';
```
이쪽 import도 손 봐주세요!

kong33 replied:
넵!!

ENvironmentSet commented on src/pages/Choosing/index.jsx:
```diff
@@ -0,0 +1,109 @@
+
+
+export const Choosing = () =>{
+  return(
```
`return` 예약어 이후에도 공백이 있으면 보기 더 좋아요.

kong33 replied:
수정했습니다!

ENvironmentSet commented on src/pages/Choosing/index.jsx:
```diff
@@ -0,0 +1,109 @@
+        position: relative;
+     
+        `}>
+        <Button buttonText='디퓨저 조향 시작하기' to='/place' backgroundColor='#2E2E2E' opacity={1}/>
```
배경색이나 투명도는 별개의 prop을 통해서 지정되기보다는 css나 className prop을 통해 지정되는 게 더 깔끔한 인터페이스가 되지 않을까요? 더해서, button text의 경우 children으로 넘기는 게 더 적절할 거 같아요.

ENvironmentSet commented on src/pages/Choosing/index.jsx:
```diff
@@ -0,0 +1,109 @@
+            font-weight: 500;
+            line-height: 50px;
+            text-transform: uppercase;
+            font-family: 'Noto Sans KR', sans-serif;
```
사용되는 문자에 따라 적절한 폰트를 자동으로 적용해 주는 컴포넌트를 만들어 보시는 건 어떨까요?

```js
<Text>
  ABC
</Text>
```

이런 식으로 사용할 수 있도록요. 혹은 전역적으로 폰트 설정을 정의해 두고 필요할 때 마다 가져다 쓸 수도 있겠어요.

```js
<p css={ko}> 안녕하세요! </p>
```

둘을 응용하면 이런 식의 컴포넌트도 만들 수 있을 거 같아요.

```js
<Text>
  My name is 재원
</Text>
```

Text 컴포넌트는 다음과 같이 렌더 트리를 생성하는 거죠.

```js
<>
  <p css={en}> My name is </p>
  <p css={ko}> 재원 </p>
</>
// inline css 주는 것은 필요하겠지만 일단 무시합시다.
```

어떤가요?

kong33 replied:
오 좋을 것 같아요 대신 english 는 폰트가 두개일텐데 그러면 prop으로 받아서 폰트 결정을 하면 될까요? 아니면 그대로 그냥 매번 적용할까요? 어떤게 더 효율적일 것 같으세요?!?

ENvironmentSet replied:
헉 이건 지금 봤네요... 이건 제가 맡아서 한번 pr 올려봐도 될까요? 제 방법을 보시고 리뷰해주시는 것도 좋을 거 같아요.

ENvironmentSet commented on src/pages/Choosing/index.jsx:
```diff
@@ -0,0 +1,109 @@
+  `}>
+    
+      <div 
+        css={css`
```
제가 작성한 다른 컴포넌트처럼 `css`의 들여쓰기를 맞춰 주세요!

ENvironmentSet commented on src/pages/Starting/index.jsx:
```diff
@@ -10,6 +10,9 @@ import blackCircle from '../../assets/black-circle.svg'
 
+import { useLocation } from 'wouter';
+
+
```
불필요한 줄바꿈을 없애주시면 좋을 거 같아요.

kong33 replied:
넵 혹시 재원님 코드에서 import 시 줄바꿈 하시는 경우가 있으시던데 그건 어떤 경우인 지 알 수 있을까요? 저는 가독성을 위해 비슷한 코드끼리 묶어서 줄바꿈 해 두 신 줄 알아서 저렇게 작성해뒀어요!

ENvironmentSet replied:
아, 이 리뷰는 줄바꿈이 두 번 들어간 걸 하나로 줄여달라는 이야기였어요!

저는 import 문을 나눌 때 다음과 같이 묶어내고는 해요.
- 라이브러리/프레임워크의 요소 (ex: `useState`, `useLocation`, `css`)
- 리소스 (ex: 위 코드의 `handUp`, `handDown`)
- 범용 유틸리티 (ex: lodash의 함수들)
- 컴포넌트

근데 이건 제가 경험적으로 터득한 기준이라 결함이 있을 수 있으니 참고만 해주시는 게 좋을 거 같아요. 관련해서도 변경을 요청드리지는 않으려고 해요~

kong33 replied:
오 저도 그렇게 나누어 코드 짜는게 더 가독성 좋은 것 같아 적용했습니다!

ENvironmentSet commented on src/pages/Starting/index.jsx:
```diff
@@ -23,15 +26,21 @@ const Background = styled.div`
 export const Starting = () => {
-  const [isTouched, setTouchStatus] = useState(false)
+  const [isTouched, setTouchStatus] = useState(false);
+  const [, setLocation] = useLocation();
```
`setLocation`(`navigate`)를 이렇게 가져오실 수도 있지만 `import { navigate } from 'wouter/use-location'`을 통해 hook을 안 거치고 가져올 수도 있습니다. [예시](https://github.com/molefrog/wouter#customizing-the-location-hook)

공식 문서에 자세한 설명이 나와있지는 않은데, 제가 코드를 열어서 살펴보니 `useLocation`에 어떤 옵션도 주지 않았을 때 얻을 수 있는 `setLocation`과 동등하게 동작하는 걸로 보였습니다.

kong33 replied:
넵 알겠습니다! 혹시 이렇게 굳이 필요 없는 코드가 있는 것 같을때는 다른 방법이 있는지 찾아보고 하시나요?! 저도 다음부터 참고하고 싶어요!!

ENvironmentSet replied:
API 문서를 찾아보시면 됩니다. 왜 이게 없지? 싶은 것들은 보통 찾아보면 다 나와요. 찾아봐도 없으면 직접 유틸리티 함수를 하나 만드시거나  그 오픈 소스에 기여하시면 됩니다(물론 이 경우에는 왜 없는지 확인하는 작업도 필요하지만요).

ENvironmentSet commented on src/pages/Starting/index.jsx:
```diff
@@ -23,15 +26,21 @@ const Background = styled.div`
     `
+  const handleTouch = () =>{
+  setTouchStatus(true);
+  setTimeout(() =>  {setLocation('/choosing');
```
`setTimeout`보다는 [`transitionend`](https://developer.mozilla.org/en-US/docs/Web/API/Element/transitionend_event) 이벤트를 사용해서 구현하는 게 더 자연스러울 거 같아요!

kong33 replied:
수정해습니다!

ENvironmentSet commented on src/pages/Starting/index.jsx:
```diff
@@ -104,7 +113,9 @@ export const Starting = () => {
       />
 
-      <p css={css`font-weight: bold; font-size: 32px; color: white;`}>화면을 터치해보세요</p>
+      <p css={css`
```
CSS property는 한 줄에 다 몰아넣으시거나 한 줄에 하나만 넣으시는 게 더 읽기 좋을 거 같아요.

kong33 replied:
넵 적용했습니다!

ENvironmentSet reviewed (changes requested):
확인 부탁드려요!

kong33 committed:
Update App.jsx

피드백 반영 - 세미콜론 삭제

kong33 committed:
Update App.jsx

피드백 반영-start page path 변경

kong33 committed:
Update index.jsx

피드백 반영 - 중괄호 안에 띄어 쓰기

kong33 committed:
button 컴포넌트의 prop에 className 추가, default css와 className css 적용 가능하게 함.

1. prop으로 className 추가
2. default css 그대로 작성 및 className이 있다면 className 우선 적용되게 함

kong33 committed:
button 컴포넌트 띄어쓰기 수정

kong33 committed:
button return에서 div 삭제

kong33 committed:
chossing 컴포넌트 띄어쓰기 수정

kong33 committed:
들여쓰기 개선, button 컴포넌트 개선에 따른 button 컴포넌트 태그 수정 

button 컴포넌트에 className prop을 추가해두었기 때문에 태그 작성 시 해당 prop 작성.

kong33 committed:
import 문 줄바꿈 삭제 - starting page

kong33 committed:
Starting 줄바꿈 수정

줄바꿈 2번 -> 1번으로 수정했습니다.

kong33 committed:
starting 페이지 settimeout 대신 transitioned 적용, navigate 직접 불러와 적용

settimeout 대신 background 태그에 id를 설정해 trasitioned를 사용해 애니메이션이 끝난 후 다음 페이지를 라우팅 했습니다. 
navigate를 wouter 라이브러리에서 직접 불러 사용했습니다.

kong33 committed:
starting 페이지 css 줄바꿈 적용

kong33 committed:
font 를 font.css로 적용

kong33 committed:
choosing 페이지 폰트 css 파일로 적용

kong33 committed:
button 컴포넌트 수정

kong33 committed:
배포 오류 수정

ENvironmentSet commented on src/App.jsx:
```diff
@@ -1,10 +1,11 @@
-    <Route path='/'> <Starting /> </Route>
-
-    <Route><Redirect to='/' /> {/** Or should we redirect to 404 page? */} </Route>
+    <Route exact path='/'> <Starting /> </Route>
```
혹시 라우팅 관련해서 어떤 이슈를 겪으셨나요? `exact`를 사용하신 이유가 있으실까요? wouter는 `exact`를 지원하지 않거든요..! (대신에 기본적으로 모든 path가 exact입니다.)

kong33 replied:
앗 exact 삭제 해둔 것 같은데 제가 되돌리기 하면서 다시 생겼나봐요!  다시 삭제해두겠습니다!!

ENvironmentSet commented on src/components/Button/index.jsx:
```diff
@@ -0,0 +1,29 @@
+import { css } from '@emotion/react'
+import '../../font.css'
```
`font.css`는 `index.html`에서 한 번만 불러오는 게 좋을 거 같아요!

kong33 replied:
넵 그런데 저희 font 파일 사용하기로 했으니 import '../../font.css'만 작성하면 될까요?? 그리고 혹시 font 파일에 있는 font가 default로 적용되고 있는 것인지 아니면 제가 style에 적용해야 하는 것인지 여쭤보고 싶어요! 지금 코드 상으로는 제가 적용해야 하는 것 같은데 혹시 default 적용 원하시는건가 해서요!

ENvironmentSet replied:
https://github.com/s-center/kiosk/blob/2ba3b6ed918f2e8d4fb8411266829d59fd3f7d88/index.html#L7

Default로 적용되고 있습니다!

kong33 replied:
아 그러면 그냥 폰트 적용 안 해도 될까요!?

ENvironmentSet replied:
네네 삭제했을 때 별 문제 없다면 그렇게 하셔도 될 거 같습니다

ENvironmentSet commented on src/pages/Starting/index.jsx:
```diff
@@ -23,15 +23,32 @@ const Background = styled.div`
+  
   return (
-    <Background onClick={() => setTouchStatus(true)}>
+    <Background  id="background" onClick={() => {setTouchStatus(true); handleTouch();}}>
```
동일한 동작을 하는 두 함수가 동시에 사용되고 있습니다. `handleTouch`가 꼭 필요할까요??

kong33 replied:
삭제하겠습니다!

ENvironmentSet commented on src/pages/Starting/index.jsx:
```diff
@@ -23,15 +23,32 @@ const Background = styled.div`
+      navigate('/choosing');
+  };
+  
+  useEffect(() => {
```
useEffect에서 dom에 접근을 해도 되기는 하지만, 그건 Refs를 사용하는 경우고 지금처럼 DOM API를 써서 직접 접근하는 건 react-dom의 오동작 및 성능상 문제를 발생시킬 수 있어 피해야 합니다. 저라면 동일한 로직을 이렇게 코딩할 거 같아요:

```jsx
<Background onClick={() => setTouchStatus(true)} onTransitionend={() => navigate('/choosinng')}>
```

ENvironmentSet replied:
추신. 다음과 같은 방법을 사용해 조건부로 prop을 추가할 수 있습니다.

```jsx
<Background onClick={...} {...condition ? { onTransitionend: () => navigate('/choosing') } : null}>
```

지금 같은 경우에는 그닥 효과적이지는 않지만요.

ENvironmentSet reviewed (changes requested):
마지막으로 몇 가지만 수정해주시면 될 거 같습니다! 고생 많으셨어요~

kong33 committed:
app.js exact 삭제

kong33 committed:
button 컴포넌트 css 중복 import 삭제

kong33 committed:
starting page choosing page로 접근 방법 변경 및 조건부 prop 추가, handleTouch, handleTransitionEnd 삭제

kong33 committed:
font css 코드 삭제

kong33 committed:
choosing 페이지 font import 삭제

kong33 committed:
font 삭제

ENvironmentSet commented:
LGTM

머지 부탁드려요~
